File: 6.18-compute_spiral_order/compute_spiral_order.py
"""
Compute the spiral order of a 2-D array

The time complexity of compute_sprial_order is O(n^2)
"""

import math

__author__ = "Matt Rathbun"
__email__ = "mrathbun80@gmail.com"
__version__ = "1.0"


def compute_sprial_order(array):
    """
    Compute the spiral order of an nxn matrix represented as a 2-D array.  This only loops through the length of the
    array / 2 because anything beyond that will have already been computed.  i.e. if the length of the array is 3 then
    it would stop at iteration 1 because if we continued the for loops of the compute_matrix_clockwise function would do
    nothing.

    i.e.
    The sprial order of
    [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]
    ]
    is [1, 2, 3, 6, 9, 8, 7, 4, 5]
    """
    result = []

    for offset in range(0, int(math.ceil(len(array) * .5))):
        compute_matrix_clockwise(array, offset, result)

    return result

def compute_matrix_clockwise(array, offset, result):
    """
    Compute the spiral order of the matrix in clockwise order.  The way this works is as follows:

    1. Add the first n-1 elements from the first row to the results list.  This happens in the first for loop.
    2. Add the first n-1 elements from the last column to the results list.  This happens in the second for loop.
    3. Add the last n-1 elements in reverse order from the last row to the results list.  This happens in the third for
       loop.
    4. Add the last n-1 elements in reverse order from the first column to the results list.  This happens in the fourth
       for loop.
    5. (Optional) If the length of the array is odd we handle the edge case where there is a middle element.

    If you were to draw a square composed of 9 smaller squares and trace your finger around the smaller squares in a
    clockwise manner starting at the upper left square and not touching any square twice it would represent what is
    happening.
    """

    # Add the middle element if the length of the array is odd
    if offset == len(array) - offset - 1:
        result.append(array[offset][offset])
        return

    # Add entries from the nth row of the array to the result
    for j in range(offset, len(array) - offset - 1):
        result.append(array[offset][j])

    # Add entries from the last column of the array to the result
    for i in range(offset, len(array) - offset - 1):
        result.append(array[i][len(array) - offset - 1])

    # Add entries from the last row of the array to the result
    for j in range(len(array) - offset - 1, offset, -1):
        result.append(array[len(array) - offset - 1][j])

    # Add entries from the nth column of the array to the result
    for i in range(len(array) - offset - 1, offset, -1):
        result.append(array[i][offset])

array = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]
array2 = [
    [1, 2],
    [3, 4]
]
array3 = [
    [1, 2, 3, 4, 5],
    [6, 7, 8, 9, 10],
    [11, 12, 13, 14, 15],
    [16, 17, 18, 19, 20],
    [21, 22, 23, 24, 25]
]
print 'The spiral order of %s is %s' % (array3, compute_sprial_order(array3))

File: php/6.18-compute_spiral_ordering/compute_spiral_order.php
<?php

require_once '../bootstrap.php';

use EOPI\Helper\InputHelper;

/**
 * The time complexity is O(n^2) and the space complexity is O(1)
 */

/**
 * Compute the spiral order of an n x n matrix.  This works by calling computeSpiralOrderClockwise for each "ring" of
 * the matrix.
 *
 * i.e.
 * If the matrix is:
 * [
 *  [1, 2, 3],
 *  [4, 5, 6],
 *  [7, 8, 9]
 * ]
 *
 * The for loop will run from 0 - 1 because ceil(0.5 * 3) is 2.  This will compute the order of the outer "ring" and
 * then handle the remaining center element.
 *
 * @param array $matrix
 * @return array
 * @throws \InvalidArgumentException
 */
function computeSpiralOrder(array $matrix)
{
    if (empty($matrix)) {
        throw new \InvalidArgumentException('The $matrix cannot be empty');
    }

    $result = [];

    for ($offset = 0; $offset < ceil(0.5 * count($matrix)); $offset++) {
        computeSpiralOrderClockwise($matrix, $offset, $result);
    }

    return $result;
}

/**
 * Compute the spiral order of an n x n matrix in a clockwise fashion.   This works by adding the first n-$offset-1
 * elements of the first row of the matrix.  Then it adds the first n-$offset-1 elements of the last column of the
 * matrix.  Then it adds the last n-$offset-1 elements of the last row, in reverse order.  Finally it adds the last
 * n-$offset-1 elements of the first column, in reverse order.  If n is odd and the $offset is equal to the center of
 * the matrix, the center element is added only.
 *
 * i.e.
 * If the matrix is:
 * [
 *  [1, 2, 3],
 *  [4, 5, 6],
 *  [7, 8, 9]
 * ]
 * And the offset is 0
 *
 * $offset = 0
 * $result = []
 *
 * Is $offset == count($matrix) - $offset - 1? No
 *
 * <<< FOR LOOP BEGIN >>>
 *
 * <<< FOR LOOP TERMINATION: $i = count($matrix) - $offset - 2 >>>
 *
 * Iteration 1:
 * $i = 0
 * $result = [1]
 *
 * Iteration 2:
 * $i = 1
 * $result = [1, 2]
 *
 * <<< FOR LOOP BEGIN >>>
 *
 * Iteration 1:
 * $j = 0
 * $result = [1, 2, 3]
 *
 * Iteration 2:
 * $j = 1
 * $result = [1, 2, 3, 6]
 *
 * <<< FOR LOOP TERMINATION: $j = count($matrix) - $offset - 2 >>>
 *
 * <<< FOR LOOP BEGIN >>>
 *
 * Iteration 1:
 * $k = 2
 * $result = [1, 2, 3, 6, 9]
 *
 * Iteration 2:
 * $k = 1
 * $result = [1, 2, 3, 6, 9, 8]
 *
 * <<< FOR LOOP TERMINATION: $k = $offset + 1 >>>
 *
 * <<< FOR LOOP BEGIN >>>
 *
 * Iteration 1:
 * $l = 2
 * $result = [1, 2, 3, 6, 9, 8, 7]
 *
 * Iteration 2:
 * $l = 1
 * $result = [1, 2, 3, 6, 9, 8, 7]
 *
 * <<< FOR LOOP TERMINATION: $l = $offset + 1 >>>
 *
 * $result = [1, 2, 3, 6, 9, 8, 7]
 *
 * =====================================================
 *
 * If the matrix is:
 * [
 *  [1, 2, 3],
 *  [4, 5, 6],
 *  [7, 8, 9]
 * ]
 * And the offset is 1
 *
 * $offset = 1
 * $result = []
 *
 * Is $offset == count($matrix) - $offset - 1? Yes
 * $result = [5]
 *
 * @param array $matrix
 * @param int $offset
 * @param array $result
 */
function computeSpiralOrderClockwise(array $matrix, $offset, array &$result)
{
    // Add the middle element
    if ($offset == count($matrix) - $offset - 1) {
        $result[] = $matrix[$offset][$offset];

        return;
    }

    // Add the n-$offset-1 elements from the first row
    for ($i = 0; $i < count($matrix) - $offset - 1; $i++) {
        $result[] = $matrix[$offset][$i];
    }

    // Add the n-$offset-1 elements of the last column
    for ($j = 0; $j < count($matrix) - $offset - 1; $j++) {
        $result[] = $matrix[$j][count($matrix) - $offset - 1];
    }

    // Add the n-$offset-1 elements of the last row, in reverse order
    for ($k = count($matrix) - $offset - 1; $k > $offset; $k--) {
        $result[] = $matrix[count($matrix) - $offset - 1][$k];
    }

    // Add the n-$offset-1 elements of the first row, in reverse order
    for ($l = count($matrix) - $offset - 1; $l > $offset; $l--) {
        $result[] = $matrix[$l][$offset];
    }
}

$inputHelper = new InputHelper();
$matrix = json_decode($inputHelper->readInputFromStdIn('Enter the 2D array representing the matrix in json format: '));
$result = computeSpiralOrder($matrix);

printf('The spiral order of %s is %s.', json_encode($matrix), json_encode($result));
print PHP_EOL;